---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# OMIM MANE-select transcript expression manuscript

> The aim with the analysis is to answer: Is MANE select/Ensembl canonical the most highly expressed transcript in the disease relevant tissue(s)?
How much of the expression is represented by MANE select/Ensembl canonical transcripts?
How much of the expression is represented by the CDS associated with MANE select?
How does this affect variant interpretation?

## Background
This repository contains all the scripts used in the OMIM MANE-select transcript expression manuscript. 

# Public long-read RNA sequencing data

**The publicly available data used comes from two sources:**

- **GTEx** - Oxford Nanopore Technologies platform (cDNA-PCR protocol) from 88 GTEx tissue and cell line samples.
- **ENCODE** - PacBio IsoSeq from 9 frontal cortex samples.


## Code contents

Within this repository you will find:

| Directory | Description |
| --------- | --------------------------------------------------------------------------- |
| [docs](docs) | Contains all `.Rmd`s and their corresponding `.html`s describing analyses performed for this project. |
| [logs](logs) | For any scripts that were run outside of an `.Rmd` (e.g. scripts from the [scripts](scripts) directory), a log file was recorded and can be accessed here. |
| [raw_data](raw_data)| Data used for the analysis. Most will not be available due to size. |
| [results](results) | Results from all analyses. |
| [scripts](scripts) | Contains analysis scripts. Each script contains a one-line description and is also referenced in its corresponding `.Rmd`. |